include/deal.II/matrix_free/tools.h

```diff
@@ -1170,99 +1832,435 @@ namespace MatrixFreeTools
+    if (face_operation)
+      data_face.op_compute = [&](auto &phi) {
+        face_operation(static_cast<FEFaceEvalType &>(*phi[0]),
+                       static_cast<FEFaceEvalType &>(*phi[1]));
+      };
 
-                local_vmult(integrator);
+    internal::ComputeMatrixScratchData<dim, VectorizedArrayType, true>
+      data_boundary;
+
+    data_boundary.dof_numbers               = {dof_no};
+    data_boundary.quad_numbers              = {dof_no};
+    data_boundary.first_selected_components = {first_selected_component};
+    data_boundary.batch_type                = {1};
+
+    data_boundary
+      .op_create = [&](const std::pair<unsigned int, unsigned int> &range) {
+      std::vector<
+        std::unique_ptr<FEEvaluationData<dim, VectorizedArrayType, true>>>
+        phi;
+
+      if (!internal::is_fe_nothing<true>(matrix_free,
+                                         range,
+                                         dof_no,
+                                         quad_no,
+                                         first_selected_component,
+                                         fe_degree,
+                                         n_q_points_1d,
+                                         true))
+        phi.emplace_back(std::make_unique<FEFaceEvalType>(
+          matrix_free, range, true, dof_no, quad_no, first_selected_component));
+
+      return phi;
+    };
+
+    data_boundary.op_reinit = [](auto &phi, const unsigned batch) {
+      static_cast<FEFaceEvalType &>(*phi[0]).reinit(batch);
+    };
+
+    if (boundary_operation)
+      data_boundary.op_compute = [&](auto &phi) {
+        boundary_operation(static_cast<FEFaceEvalType &>(*phi[0]));
+      };
+
+    internal::compute_matrix(
+      matrix_free, constraints_in, data_cell, data_face, data_boundary, matrix);
+  }
+
+  namespace internal
+  {
+    template <int dim,
+              typename Number,
+              typename VectorizedArrayType,
+              typename MatrixType>
+    void
+    compute_matrix(
+      const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+      const AffineConstraints<Number>                    &constraints_in,
+      const internal::ComputeMatrixScratchData<dim, VectorizedArrayType, false>
+        &data_cell,
+      const internal::ComputeMatrixScratchData<dim, VectorizedArrayType, true>
+        &data_face,
+      const internal::ComputeMatrixScratchData<dim, VectorizedArrayType, true>
+                 &data_boundary,
+      MatrixType &matrix)
+    {
+      std::unique_ptr<AffineConstraints<typename MatrixType::value_type>>
+        constraints_for_matrix;
+      const AffineConstraints<typename MatrixType::value_type> &constraints =
+        internal::create_new_affine_constraints_if_needed(
+          matrix, constraints_in, constraints_for_matrix);
+
+      const auto batch_operation =
+        [&matrix_free, &constraints, &matrix](
+          auto &data, const std::pair<unsigned int, unsigned int> &range) {
```

Comment:
Can you specify the type of `data` here?

Reply:
This is not possible. It is either `internal::ComputeMatrixScratchData<dim, VectorizedArrayType, false>` or `internal::ComputeMatrixScratchData<dim, VectorizedArrayType, true>`.

Reply:
OK I see, let us keep it.